man/cw-mesh.1.pod
  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.


=head1 NAME

cw-mesh - Clockwork Mesh Client

=head1 SYNOPSIS

B<cw-mesh> [OPTIONS] command [arg...]

=head1 DESCRIPTION

#INTRO

B<cw-mesh> is the client utility for Clockwork Mesh.  With it you can
interact with remote Clockwork nodes, query them for resource state (what
version of Clockwork are you running?  Is package X installd?) start
services, execute commands, etc.

You can read more about Mesh in the B<mesh>(7) man page.

There are several alias commands to make using mesh more natural.

    $ cw show [OPTIONS] version

is identical to

    $ cw mesh [OPTIONS] show version

The former makes it more natural to I<speak> mesh, especially if there are
no options to be had.

=head1 OPTIONS

=over

=item B<-t>, B<--timeout> I<seconds>

How long to wait for responses from mesh nodes, in seconds.  B<cw-mesh> will
B<always> wait this long, since there is no reliable way to determine how
many nodes will respond to the query.

Defaults to I<40> seconds.  Values of 0 or less will be treated as this
default.

This option can be set in the global I</etc/clockwork/cw.conf> and
overridden in B<~/.cwrc>, via the B<mesh.timeout> directive.
The command-line flag will override both of those values.

=item B<-s>, B<--sleep> I<milliseconds>

How many milliseconds to sleep between checking for results.

Defaults to I<250> milliseconds.  The minimum value is I<100> milliseconds.

This option can be set in the global I</etc/clockwork/cw.conf> and
overridden in B<~/.cwrc>, via the B<mesh.sleep> directive.
The command-line flag will override both of those values.

=item B<-w>, B<--where> I<FILTER>

The given command should only be run on nodes matching I<FILTER>.  Filters
are specified as 'key op value'.  Regex matching is supported.

Here are the valid forms for a filter:

    fact.name =  literal   # match the value "literal"
    fact.name != literal   # match any other value

    fact.name =  /^web/    # match any value starting with "web"
    fact.name != /^web/    # match any value not starting with "web"

Note that '=' and '!=' are used for both literal matches and regex matches;
the magic is in the forward slashes that surround the regex pattern.

    fact.name = "^web"     # match the literal value "^web"

Multiple filters can be givne by adding the B<-w> flag multiple times.  All
filters are ANDed together (they all have to match).

=item B<-v>, B<--verbose>

=item B<-q>, B<--quiet>

How much should B<cw-mesh> print out.  This is primarily for diagnostics.

=item B<-n>, B<--noop>

Parse options and the command, to ensure they are all valid, but don't
authenticate or submit the query.

=item B<-u>, B<--username> I<USERNAME>

User to authenticate as.  If B<cw-mesh> can't figure out who the current
user is, and this option is not specified, it will prompt for the username.

=item B<-k>, B<--key> I</path/to/mesh.key>

Use the specified user private key and do key-based authentication.  No
passwords will be prompted for, and password authentication will be skipped.

Defaults to I<~/clockwork/mesh.key>.

If the special value "0" is given (as in B<-k0>), then public key
authentication will be skipped, and you B<will> be prompted for a password.

=item B<--cert> I</path/to/cert.pub>

Path to the public certificate for validating the Mesh master server.

There is no default for this option, but it can (and should) be set by the
global I</etc/clockwork/cw.conf> or B<~/.cwrc> via the B<mesh.cert>
directive.

=item B<-c>, B<--config> I</path/to/cw.conf>

Specify an alternate B<cw> configuration file.  If not given, B<cw-mesh>
will check the normal places, starting with the global
I</etc/clockwork/cw.conf> file, followed by overrides in B<~/.cwrc>.
If you do specify a config file, the normal files (including B<~/.cwrc>)
will be summarily skipped.

=item B<--optouts>

By default, B<cw-mesh> will ignore all B<OPTOUT> replies from nodes that
received the command, but opted out due to filtering and applicability.  For
debugging and troubleshooting, it can be helpful to see these replies, to
verify that a missing host is in fact not responding at all, rather than
mistakenly (or unexpectedly) opting out.

=item B<-Y>, B<--yaml>

Output results in YAML format.  See B<OUTPUT FORMATS> for examples.

=item B<-J>, B<--json>

Output results in JSON format.  See B<OUTPUT FORMATS> for examples.

=item B<-X>, B<--xml>

Output results in XML format.  See B<OUTPUT FORMATS> for examples.

=item B<-T>, B<--text>

Output results in a simple, machine-readable text format.  This format is
intended to be easily parsed by simple shell scripts.

See B<OUTPUT FORMATS> for examples.

=item B<-P>, B<--pretty>

Output results in a pretty, human-readable format.  This is the default.

See B<OUTPUT FORMATS> for examples.

=back

=head1 OUTPUT FORMATS

=head2 YAML

    ---
    fe8565d6-2c2b-43e6-9c84-58a57f7d33ee:
      fqdn: host1.example.com
      status: 0
      output: |-
        command output
    aef5002e-9238-43fe-b280-3e197136bb53:
      fqdn: host2.example.com
      status: 3
      output: |-
        more output
        on multiple lines
    d79835c2-7e0f-41c4-adb7-f87596229542:
      fqdn: optout.example.com
      status: ~
      output: ~

=head2 JSON (--json)

    {
      "fe8565d6-2c2b-43e6-9c84-58a57f7d33ee": {
        "fqdn"   : "host1.example.com",
        "status" : 0,
        "output" : "command output\n"
      },
      "aef5002e-9238-43fe-b280-3e197136bb53": {
        "fqdn"   : "host2.example.com",
        "status" : 3,
        "output" : "more output\non multiple lines\n"
      },
      "d79835c2-7e0f-41c4-adb7-f87596229542": {
        "fqdn"   : "optout.example.com",
        "status" : null,
        "output" : null
      },
    }

=head2 XML (--xml)

    <?xml version="1.0" encoding="UTF-8"?>
    <results>
      <result uuid="fe8565d6-2c2b-43e6-9c84-58a57f7d33ee">
        <fqdn>host1.example.com</fqdn>
        <status>0</status>
        <output><![CDATA[command output
    </output>
      </result>
      <result uuid="aef5002e-9238-43fe-b280-3e197136bb53">
        <fqdn>host2.example.com</fqdn>
        <status>3</status>
        <output><![CDATA[more output
    on multiple lines
    ]]></output>
      </result>
      <result uuid="d79835c2-7e0f-41c4-adb7-f87596229542">
        <fqdn>optout.example.com</fqdn>
        <optout/>
      </result>
    </results>

=head2 Machine-parseable Text (--text)

    fe8565d6-2c2b-43e6-9c84-58a57f7d33ee host1.example.com success 0
     command output
    aef5002e-9238-43fe-b280-3e197136bb53 host2.example.com failed 3
     more output
     on multiple lines
    d79835c2-7e0f-41c4-adb7-f87596229542 optout.example.com optout

=head2 Human-readable Text (--pretty)

    host1.example.com succeeded
    UUID fe8565d6-2c2b-43e6-9c84-58a57f7d33ee
    ------------------------------------------------------------
      command output


    host2.example.com failed (rc=3)
    UUID aef5002e-9238-43fe-b280-3e197136bb53
    ------------------------------------------------------------
     more output
     on multiple lines


    optout.example.com opted out
    UUID d79835c2-7e0f-41c4-adb7-f87596229542

=head1 AUTHOR

#AUTHOR

=cut


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




cli/cw-mesh.c
/*
  Copyright 2011-2014 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <termios.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sodium.h>

#include "../src/clockwork.h"
#include "../src/mesh.h"

#define PROMPT_ECHO   1
#define PROMPT_NOECHO 0

#define FORMAT_YAML   1
#define FORMAT_JSON   2
#define FORMAT_XML    3
#define FORMAT_TEXT   4
#define FORMAT_PRETTY 5

static int SHOW_OPTOUTS = 0;
static int FORMAT = FORMAT_PRETTY;

static char* s_prompt(const char *prompt, int echo)
{
	struct termios term_new, term_old;
	if (!echo) {
		if (tcgetattr(0, &term_old) != 0)
			return NULL;
		term_new = term_old;
		term_new.c_lflag &= ~ECHO;

		if (tcsetattr(0, TCSAFLUSH, &term_new) != 0)
			return NULL;
	}

	int attempts = 0;
	char answer[8192] = {0};
	while (!*answer) {
		if (attempts++ > 1) fprintf(stderr, "\n");
		fprintf(stderr, "%s", prompt);
		if (!fgets(answer, 8192, stdin))
			answer[0] = '\0';

		char *newline = strchr(answer, '\n');
		if (newline) *newline = '\0';
	}

	if (!echo) {
		tcsetattr(0, TCSAFLUSH, &term_old);
		fprintf(stderr, "\n");
	}

	return strdup(answer);
}

typedef struct {
	char      *endpoint;
	char      *username;
	char      *password;
	char      *authkey;
	cmd_t     *command;
	char      *filters;

	int        timeout_ms;
	int        sleep_ms;

	cw_cert_t *client_cert;
	cw_cert_t *master_cert;
	cw_cert_t *auth_cert;
} client_t;

static client_t* s_client_new(void)
{
	client_t *c = cw_alloc(sizeof(client_t));

	c->client_cert = cw_cert_generate(CW_CERT_TYPE_ENCRYPTION);

	c->username = getenv("USER");
	if (c->username) c->username = strdup(c->username);

	if (getenv("HOME")) {
		c->authkey = cw_string("%s/.clockwork/mesh.key", getenv("HOME"));

		struct stat st;
		if (stat(c->authkey, &st) != 0) {
			cw_log(LOG_INFO, "Default key %s not found; skipping implicit keyauth", c->authkey);
			free(c->authkey);
			c->authkey = NULL;
		}
	}

	return c;
}

static void s_client_free(client_t *c)
{
	if (c) return;

	cw_cert_destroy(c->client_cert);
	cw_cert_destroy(c->master_cert);

	free(c->filters);

	free(c->endpoint);
	free(c->username);
	free(c->password);

	free(c);
}

static int s_read_config(cw_list_t *cfg, const char *file, int must_exist)
{
	assert(file);
	assert(cfg);

	struct stat st;
	if (stat(file, &st) != 0) {
		if (must_exist) {
			fprintf(stderr, "%s: %s\n", file, strerror(ENOENT));
			return 1;
		}
		return 0;
	}

	FILE *io = fopen(file, "r");
	if (!io) {
		fprintf(stderr, "%s: %s\n", file, strerror(errno));
		return 1;
	}

	if (cw_cfg_read(cfg, io) != 0) {
		fprintf(stderr, "Unable to grok %s\n", file);
		fclose(io);
		return 1;
	}

	fclose(io);
	return 0;
}

static client_t* s_init(int argc, char **argv)
{
	client_t *c = s_client_new();

	const char *short_opts = "+h?Vvqnu:k:t:s:w:c:YJXTP";
	struct option long_opts[] = {
		{ "help",        no_argument,       NULL, 'h' },
		{ "version",     no_argument,       NULL, 'V' },
		{ "verbose",     no_argument,       NULL, 'v' },
		{ "quiet",       no_argument,       NULL, 'q' },
		{ "noop",        no_argument,       NULL, 'n' },
		{ "username",    required_argument, NULL, 'u' },
		{ "key",         required_argument, NULL, 'k' },
		{ "timeout",     required_argument, NULL, 't' },
		{ "sleep",       required_argument, NULL, 's' },
		{ "where",       required_argument, NULL, 'w' },
		{ "config",      required_argument, NULL, 'c' },
		{ "optouts",     no_argument,       NULL,  1  },
		{ "cert",        required_argument, NULL,  2  },

		{ "yaml",        no_argument,       NULL, 'Y' },
		{ "json",        no_argument,       NULL, 'J' },
		{ "xml",         no_argument,       NULL, 'X' },
		{ "text",        no_argument,       NULL, 'T' },
		{ "pretty",      no_argument,       NULL, 'P' },
		{ 0, 0, 0, 0 },
	};
	int verbose = LOG_WARNING, noop = 0;
	cw_strl_t *filters = cw_strl_new(NULL);

	cw_log_open("cw-mesh", "stderr");
	cw_log_level(LOG_ERR, NULL);

	char *x, *config_file = NULL;
	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("cw-mesh, part of clockwork v%s\n", PACKAGE_VERSION);
			exit(0);

		case 'V':
			printf("cw-mesh (Clockwork) %s\n"
			       "Copyright (C) 2014 James Hunt\n",
			       PACKAGE_VERSION);
			exit(0);

		case 'v':
			verbose++;
			if (verbose > LOG_DEBUG) verbose = LOG_DEBUG;
			cw_log_level(verbose, NULL);
			break;

		case 'q':
			verbose = LOG_WARNING;
			cw_log_level(verbose, NULL);
			break;

		case 'n':
			noop = 1;
			break;

		case 'u':
			free(c->username);
			c->username = strdup(optarg);
			break;

		case 'k':
			free(c->authkey);
			c->authkey = strcmp(optarg, "0") == 0 ? NULL : strdup(optarg);
			break;

		case 't':
			c->timeout_ms = atoi(optarg) * 1000;
			if (c->timeout_ms <=    0) c->timeout_ms = 40 * 1000;
			if (c->timeout_ms <  1000) c->timeout_ms =  1 * 1000;
			break;

		case 's':
			c->sleep_ms = atoi(optarg);
			if (c->sleep_ms <=   0) c->sleep_ms = 250;
			if (c->sleep_ms <  100) c->sleep_ms = 100;
			break;

		case 'w':
			x = cw_string("%s\n", optarg);
			cw_strl_add(filters, x);
			free(x);
			break;

		case 'c':
			free(config_file);
			config_file = strdup(optarg);
			break;

		case 1:
			SHOW_OPTOUTS = 1;
			break;

		case 2:
			cw_cert_destroy(c->master_cert);
			c->master_cert = cw_cert_read(optarg);
			if (!c->master_cert) {
				fprintf(stderr, "%s: %s", optarg, errno == EINVAL
					? "Invalid Clockwork certificate"
					: strerror(errno));
				exit(1);
			}
			break;

		case 'Y': FORMAT = FORMAT_YAML;   break;
		case 'J': FORMAT = FORMAT_JSON;   break;
		case 'X': FORMAT = FORMAT_XML;    break;
		case 'T': FORMAT = FORMAT_TEXT;   break;
		case 'P': FORMAT = FORMAT_PRETTY; break;

		default:
			fprintf(stderr, "Unknown flag '-%c' (optind %i / optarg '%s')\n", opt, optind, optarg);
			exit(1);
		}
	}

	cw_log_level(verbose, NULL);
	c->filters = cw_strl_join(filters, "");

	LIST(config);
	cw_cfg_set(&config, "mesh.timeout",  "40");
	cw_cfg_set(&config, "mesh.sleep",   "250");

	if (!config_file) {
		if (s_read_config(&config, CW_MTOOL_CONFIG_FILE, 0) != 0)
			exit(1);

		if (getenv("HOME")) {
			char *path = cw_string("%s/.cwrc", getenv("HOME"));
			if (s_read_config(&config, path, 0) != 0)
				exit(1);

			config_file = cw_string("%s or %s", CW_MTOOL_CONFIG_FILE, path);
			free(path);

		} else {
			config_file = strdup(CW_MTOOL_CONFIG_FILE);
		}

	} else if (s_read_config(&config, config_file, 1) != 0) {
		exit(1);
	}

	if (c->timeout_ms == 0)
		c->timeout_ms = atoi(cw_cfg_get(&config, "mesh.timeout")) * 1000;
	if (c->timeout_ms < 1000)
		c->timeout_ms = 1000;

	if (c->sleep_ms == 0)
		c->sleep_ms = atoi(cw_cfg_get(&config, "mesh.sleep"));
	if (c->sleep_ms < 100)
		c->sleep_ms = 100;

	c->endpoint = cw_cfg_get(&config, "mesh.master");
	if (!c->endpoint) {
		fprintf(stderr, "mesh.master not found in %s\n", config_file);
		exit(1);
	}
	c->endpoint = cw_string("tcp://%s", c->endpoint);

	if (!c->master_cert) {
		if (!cw_cfg_get(&config, "mesh.cert")) {
			fprintf(stderr, "mesh.cert not found in %s, and --cert not specified\n",
				config_file);
			exit(1);
		}

		c->master_cert = cw_cert_read(cw_cfg_get(&config, "mesh.cert"));
		if (!c->master_cert) {
			fprintf(stderr, "%s: %s\n", cw_cfg_get(&config, "mesh.cert"),
				errno == EINVAL ? "Invalid Clockwork certificate" : strerror(errno));
			exit(1);
		}
		if (!c->master_cert->pubkey) {
			fprintf(stderr, "%s: no public key component found\n",
				cw_cfg_get(&config, "mesh.cert"));
			exit(1);
		}
	}

	if (c->authkey) {
		c->auth_cert = cw_cert_read(c->authkey);
		if (!c->auth_cert) {
			fprintf(stderr, "%s: %s\n", c->authkey,
				errno == EINVAL ? "Invalid Clockwork key" : strerror(errno));
			exit(1);
		}
		if (c->auth_cert->type != CW_CERT_TYPE_SIGNING) {
			fprintf(stderr, "%s: Incorrect key/certificate type (not a %%sig v1 key)\n", c->authkey);
			exit(1);
		}
		if (!c->auth_cert->seckey) {
			fprintf(stderr, "%s: no secret key component found\n", c->authkey);
			exit(1);
		}
	}

	cw_log(LOG_DEBUG, "Parsing command from what's left of argv");
	cmd_destroy(c->command);
	c->command = cmd_parsev((const char **)argv + optind, COMMAND_LITERAL);
	if (!c->command) {
		cw_log(LOG_ERR, "Failed to parse command!");
		exit(2);
	}

	if (noop) exit(0);

	cw_log(LOG_INFO, "Running command `%s`", c->command->string);

	return c;
}

void print_header(void)
{
	switch (FORMAT) {
	case FORMAT_YAML:
		printf("---\n");
		break;

	case FORMAT_JSON:
		printf("{\n");
		break;

	case FORMAT_XML:
		printf("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
		printf("<results>\n");
		break;
	}
}

void print_optout(const char *fqdn, const char *uuid)
{
	switch (FORMAT) {
	case FORMAT_YAML:
		printf("%s:\n", uuid);
		printf("  fqdn: %s\n", fqdn);
		printf("  status: ~\n");
		printf("  output: ~\n");
		break;

	case FORMAT_JSON:
		printf("  \"%s\": {\n", uuid);
		printf("    \"fqdn\"   : \"%s\",\n", fqdn);
		printf("    \"status\" : null,\n");
		printf("    \"output\" : null\n");
		printf("  },\n");
		break;

	case FORMAT_XML:
		printf("  <result uuid=\"%s\">\n", uuid);
		printf("    <fqdn>%s</fqdn>\n", fqdn);
		printf("    <optout/>\n");
		printf("  </result>\n");
		break;

	case FORMAT_TEXT:
		printf("%s %s optout\n", uuid, fqdn);
		break;

	case FORMAT_PRETTY:
		printf("%s opted out\nUUID %s\n\n", fqdn, uuid);
		break;
	}
}

static void printf_eachline(const char *fmt, const char *buf)
{
	size_t len = 0;
	const char *a, *b;

	for (a = buf, b = strchr(a, '\n'); b; a = b + 1, b = strchr(a, '\n'))
		if (b - a > len)
			len = b - a;

	char *tmp = cw_alloc(len + 1);
	for (a = buf, b = strchr(a, '\n'); b; a = b + 1, b = strchr(a, '\n')) {
		memcpy(tmp, a, b - a);
		tmp[b-a] = '\0';
		printf(fmt, tmp);
	}
}

void print_result(const char *fqdn, const char *uuid, const char *result, const char *output)
{
	switch (FORMAT) {
	case FORMAT_YAML:
		printf("%s:\n", uuid);
		printf("  fqdn: %s\n", fqdn);
		printf("  status: %s\n", result);
		printf("  output: |-\n");
		printf_eachline("    %s\n", output);
		break;

	case FORMAT_JSON:
		printf("  \"%s\": {\n", uuid);
		printf("    \"fqdn\"   : \"%s\",\n", fqdn);
		printf("    \"status\" : \"%s\",\n", result);
		printf("    \"output\" : \"%s\"\n",  output); /* FIXME: escape quotes/newlines! */
		printf("  }\n");
		break;

	case FORMAT_XML:
		printf("  <result uuid=\"%s\">\n", uuid);
		printf("    <fqdn>%s</fqdn>\n", fqdn);
		printf("    <status>%s</status>\n", result);
		printf("    <output><![CDATA[%s]]></output>\n", output); /* FIXME: escape ']]>' */
		printf("  </result>\n");
		break;

	case FORMAT_TEXT:
		printf("%s %s %s %s\n", uuid, fqdn,
			strcmp(result, "0") == 0 ? "success" : "failed",
			result);
		printf_eachline(" %s\n", output);
		break;

	case FORMAT_PRETTY:
		if (strcmp(result, "0") == 0)
			printf("%s succeeded\n", fqdn);
		else
			printf("%s failed (rc=%s)\n", fqdn, result);
		printf("UUID %s\n", uuid);
		printf("------------------------------------------------------------\n");
		printf_eachline("  %s\n", output);
		printf("\n\n");
		break;
	}
}

void print_footer(void)
{
	switch (FORMAT) {
	case FORMAT_JSON:
		printf("}\n");
		break;

	case FORMAT_XML:
		printf("</results>\n");
		break;
	}
}

int main(int argc, char **argv)
{
	client_t *c = s_init(argc, argv);
	int rc;

	cw_log(LOG_DEBUG, "Setting up 0MQ context");
	void *zmq = zmq_ctx_new();
	void *zap = cw_zap_startup(zmq, NULL);

	void *client = zmq_socket(zmq, ZMQ_DEALER);

	rc = zmq_setsockopt(client, ZMQ_CURVE_SECRETKEY, cw_cert_secret(c->client_cert), 32);
	assert(rc == 0);
	rc = zmq_setsockopt(client, ZMQ_CURVE_PUBLICKEY, cw_cert_public(c->client_cert), 32);
	assert(rc == 0);
	rc = zmq_setsockopt(client, ZMQ_CURVE_SERVERKEY, cw_cert_public(c->master_cert), 32);
	assert(rc == 0);
	cw_log(LOG_INFO, "Connecting to %s", c->endpoint);
	rc = zmq_connect(client, c->endpoint);
	assert(rc == 0);


	if (!c->username)
		c->username = s_prompt("Username: ", PROMPT_ECHO);
	if (!c->auth_cert)
		c->password = s_prompt("Password: ", PROMPT_NOECHO);

	cw_pdu_t *pdu = cw_pdu_make(NULL, 2, "REQUEST", c->username);
	if (c->auth_cert) {
		assert(c->auth_cert);

		unsigned char unsealed[256 - 64];
		randombytes(unsealed, 256 - 64);

		uint8_t *sealed;
		unsigned long long slen;
		slen = cw_cert_seal(c->auth_cert, unsealed, 256 - 64, &sealed);
		assert(slen == 256);

		char *pubkey = cw_cert_public_s(c->auth_cert);
		cw_pdu_extend(pdu, cw_frame_new(pubkey));
		cw_pdu_extend(pdu, cw_frame_newbuf((const char*)sealed, 256));

		free(pubkey);
		free(sealed);
	} else {
		assert(c->password);
		cw_pdu_extend(pdu, cw_frame_new(""));
		cw_pdu_extend(pdu, cw_frame_new(c->password));
	}
	cw_pdu_extend(pdu, cw_frame_new(c->command->string));
	cw_pdu_extend(pdu, cw_frame_new(c->filters));

	rc = cw_pdu_send(client, pdu);
	assert(rc == 0);

	cw_pdu_t *reply = cw_pdu_recv(client);
	if (strcmp(reply->type, "ERROR") == 0) {
		cw_log(LOG_ERR, "ERROR: %s", cw_pdu_text(reply, 1));
		exit(3);
	}
	if (strcmp(reply->type, "SUBMITTED") != 0) {
		cw_log(LOG_ERR, "Unknown response: %s", reply->type);
		exit(3);
	}

	char *serial = cw_pdu_text(reply, 1);
	cw_pdu_destroy(reply);
	cw_log(LOG_INFO, "query submitted; serial = %s", serial);

	print_header();

	int n;
	for (n = c->timeout_ms / c->sleep_ms; n > 0; n--) {
		cw_sleep_ms(c->sleep_ms);

		cw_log(LOG_INFO, "awaiting result, countdown = %i", n);
		pdu = cw_pdu_make(NULL, 2, "CHECK", serial);
		rc = cw_pdu_send(client, pdu);
		assert(rc == 0);
		cw_pdu_destroy(pdu);

		for (;;) {
			reply = cw_pdu_recv(client);
			if (strcmp(reply->type, "DONE") == 0) {
				cw_pdu_destroy(reply);
				break;
			}

			if (strcmp(reply->type, "ERROR") == 0) {
				cw_log(LOG_ERR, "ERROR: %s", cw_pdu_text(reply, 1));
				cw_pdu_destroy(reply);
				break;
			}


			if (strcmp(reply->type, "OPTOUT") == 0) {
				if (SHOW_OPTOUTS) {
					char *fqdn = cw_pdu_text(reply, 1);
					char *uuid = cw_pdu_text(reply, 2);
					print_optout(fqdn, uuid);

					free(fqdn);
					free(uuid);
				}

			} else if (strcmp(reply->type, "RESULT") == 0) {
				char *fqdn   = cw_pdu_text(reply, 1);
				char *uuid   = cw_pdu_text(reply, 2);
				char *result = cw_pdu_text(reply, 3);
				char *output = cw_pdu_text(reply, 4);

				print_result(fqdn, uuid, result, output);

				free(fqdn);
				free(uuid);
				free(result);
				free(output);

			} else {
				cw_log(LOG_ERR, "Unknown response: %s", reply->type);
				cw_pdu_destroy(reply);
				break;
			}

			cw_pdu_destroy(reply);
		}
	}

	print_footer();

	cw_log(LOG_INFO, "cw-mesh shutting down");


	cw_zmq_shutdown(client, 500);
	cw_zap_shutdown(zap);
	cw_cert_destroy(c->client_cert);
	cw_cert_destroy(c->master_cert);
	s_client_free(c);
	return 0;
}
